miden-lib/asm/sat/internal/account_vault.masm

```diff
@@ -63,3 +66,227 @@ export.has_non_fungible_asset
+#! Output: [ASSET]
+#!
+#! - ASSET is the fungible asset to remove from the vault.
+proc.remove_fungible_asset
+    dupw push.0 movdn.3
+    # => [ASSET_KEY, amount, ASSET]
+
+    # get the asset vault root and read asset
+    exec.layout::get_acct_vault_root swapw exec.smt::get
+    # => [CUR_ASSET, VAULT_ROOT, amount, ASSET]
+
+    # arrange elements
+    movup.3 movup.8 dup dup.3
+    # => [cur_amount, amount, amount, cur_amount, faucet_id, 0, 0, VAULT_ROOT, ASSET]
+
+    # assert amount <= cur_amount
+    lte assert
+    # => [amount, cur_amount, faucet_id, 0, 0, VAULT_ROOT, ASSET]
+
+    # asset amount + cur_amount < max_amount
+    sub
+    # => [new_amount, faucet_id, 0, 0, VAULT_ROOT, ASSET]
+
+    # => check if the asset amount is zero
+    dup eq.0
+    # => [is_zero, new_amount, faucet_id, 0, 0, VAULT_ROOT, ASSET]
+
+    if.true
+        # fungible asset empty - insert EMPTY_WORD in vault
+        movdn.3 padw
+        # => [EMPTY_WORD, ASSET_KEY, VAULT_ROOT, ASSET]
+    else
+        # fungible asset not empty - update asset in vault
+        movdn.3 dupw movup.3 drop push.0 movdn.3 swapw
+        # => [NEW_ASSET, ASSET_KEY, VAULT_ROOT, ASSET]
+    end
+
+    # update asset in vault
+    exec.smt::insert dropw
+    # => [VAULT_ROOT', ASSET]
+
+    # update the vault root
+    exec.layout::set_acct_vault_root
+    # => [ASSET]
+end
+
+#! Remove the specified non-fungible asset from the vault.
+#!
+#! Panics:
+#! - The non-fungible asset is not found in the vault.
+#!
+#! Stack: [ASSET]
+#! Output: [ASSET]
+#!
+#! - ASSET is the non-fungible asset to remove from the vault.
+proc.remove_non_fungible_asset
+    # prepare the stack to insert an EMPTY_WORD into the vault at key associated with the 
+    # non-fungible asset
+    dupw exec.layout::get_acct_vault_root swapw padw
+    # => [EMPTY_WORD, ASSET, VAULT_ROOT, ASSET]
+
+    # update asset in vault
+    exec.smt::insert
+    # => [OLD_VAL, VAULT_ROOT', ASSET]
+
+    # Assert old value was not empty (we only need to check ASSET[1] which is the faucet id)
+    drop drop eq.0 assertz drop
+    # => [VAULT_ROOT', ASSET]
+
+    # update the vault root
+    exec.layout::set_acct_vault_root
+    # => [ASSET]
+end
+
+#! Remove the specified asset from the vault.
+#!
+#! Panics:
+#! - The fungible asset is not found in the vault.
+#! - The amount of the fungible asset in the vault is less than the amount to be removed.
+#! - The non-fungible asset is not found in the vault.
+#!
+#! Stack: [ASSET]
+#! Output: [ASSET]
+#!
+#! - ASSET is the asset to remove from the vault.
+export.remove_asset
+    # assert that the asset is valid
+    exec.asset::validate_asset
+    # => [ASSET]
+
+    # check if the asset is a fungible asset
+    exec.asset::is_fungible_asset
+    # => [is_fungible_asset, ASSET]
+
+    # add the asset to the asset vault
+    if.true
+        exec.remove_fungible_asset
+        # => [ASSET]
+    else
+        exec.remove_non_fungible_asset
+        # => [ASSET]
+    end
+end
```

Comment:
Similar comment to the one above - but also, in this case, if an invalid asset is provided, we would not have it in the vault. So, the check for asset validity is implied in this, right?

Reply:
good catch.  Lets remove the asset validation here.